feat(backup-reports): mjml template

b-Nollet opened the pull request:
### Description

Backup reports using MJML.

![image](https://github.com/vatesfr/xen-orchestra/assets/150058367/d7223188-794a-40d7-82b1-77bf58af9f46)

![image](https://github.com/vatesfr/xen-orchestra/assets/150058367/480dfa80-bcb8-433d-9ec0-f2dab4179ca4)
![image](https://github.com/vatesfr/xen-orchestra/assets/150058367/c62809ff-ff85-4636-83f9-e312f12d7391)


### Checklist

- Commit
  - Title follows [commit conventions](https://bit.ly/commit-conventions)
  - Reference the relevant issue (`Fixes #007`, `See xoa-support#42`, `See https://...`)
  - If bug fix, add `Introduced by`
- Changelog
  - If visible by XOA users, add changelog entry
  - Update "Packages to release" in `CHANGELOG.unreleased.md`
- PR
  - If UI changes, add screenshots
  - If not finished or not tested, open as _Draft_

b-Nollet committed:
feat(backups-reports): mjml backup reports

b-Nollet committed:
first version

julien-f commented on packages/xo-server-backup-reports/src/index.js:
```diff
@@ -52,27 +53,38 @@ export const testSchema = {
-const compiledVmTemplate = Handlebars.compile(
-  readFileSync(join(__dirname, '../templates/vm.hbs')).toString().replace(/\n$/, '')
-)
+registerPartials('../templates/markdown/partials/')
```
Use separate Handlebars instances, to avoid conflicts between templates.

julien-f commented on packages/xo-server-backup-reports/src/index.js:
```diff
@@ -206,9 +218,16 @@ class BackupReportsXoPlugin {
-      markdown: compiledMetadataTemplate(context),
+      subject: templates.subject.metadata(context),
+      markdown: templates.markdown.metadata(context),
+      html: mjml2html(templates.mjml.metadata(context)).html,
```
Move this logic in a template transform:

```js
export default async function transform (source) {
  return { html: Mjml.compile(source) }
}
```

b-Nollet committed:
templates improvements

- added template transform
- separated Handlebars environments
- improved template look

b-Nollet committed:
removed hard-coded markdown link

b-Nollet committed:
yarn.lock

b-Nollet committed:
changelog

julien-f commented on packages/xo-server-backup-reports/src/helpers.js:
```diff
@@ -15,15 +14,15 @@ const STATUS_ICON = {
 // ===================================================================
 
-Handlebars.registerHelper('ifCond', function (v1, operator, v2, options) {
+const ifCond = function (v1, operator, v2, options) {
```
```suggestion
function ifCond(v1, operator, v2, options) {
```

julien-f commented on packages/xo-server-backup-reports/src/helpers.js:
```diff
@@ -44,36 +43,47 @@ Handlebars.registerHelper('ifCond', function (v1, operator, v2, options) {
+}
 
-Handlebars.registerHelper('formatDuration', milliseconds => moment.duration(milliseconds).humanize())
+const formatDuration = milliseconds => moment.duration(milliseconds).humanize()
```
Export directly helpers:

```js
export const formatDuration
```

julien-f commented on packages/xo-server-backup-reports/src/index.js:
```diff
@@ -1,12 +1,14 @@
 import { forEach, groupBy } from 'lodash'
 import { get } from '@xen-orchestra/defined'
-import { extname, join, parse } from 'node:path'
+import { helpers } from './helpers'
```
```suggestion
import * as helpers from './helpers'
```

julien-f commented on packages/xo-server-backup-reports/src/index.js:
```diff
@@ -1,12 +1,14 @@
 import Handlebars from 'handlebars'
 import moment from 'moment-timezone'
+import pkg from '../package'
```
I like to group exports:

- absolute
- parent dir
- current dir

```js
import foo from 'foo'
import qux from 'qux'

import bar from '../bar'

import baz from './baz'
```

julien-f commented on packages/xo-server-backup-reports/src/index.js:
```diff
@@ -1,12 +1,14 @@
-import pkg from '../package'
-import './helpers'
+import markdownTransform from '../templates/markdown/transform.js'
+import mjmlTransform from '../templates/mjml/transform.js'
```
**low priority**

Import them "dynamically":

```js
const templates = new Map()
for (const name of ['markdown', 'mjml']) {
  // todo
}
```

julien-f commented on packages/xo-server-backup-reports/src/index.js:
```diff
@@ -207,8 +224,9 @@ class BackupReportsXoPlugin {
+      html: mjmlTransform.transform(templates.mjml.metadata(context)),
+      subject: templates.mjml.metadataSubject(context),
       success: log.status === 'success',
     })
```
This uncouples the templates from the plugin code even further.

```js
return this._sendReport({
  ...await templates.markdown.transform(template.markdown.metadata(context)),
  ...await templates.mjml.transform(template.mjml.metadata(context)),
  subject: templates.mjml.metadataSubject(context),
  success: log.status === 'success',
})
```

b-Nollet committed:
suggested changes

b-Nollet committed:
dynamic transform import

julien-f committed:
refresh yarn.lock

julien-f committed:
fix incorrect rebase
